fix: replace regex usage with simple search

strehle opened the pull request:
Refactor use of regex to create a count statement
Issue found by Sonar

strehle committed:
fix regex usage

Refactor use of regex to create a count statement
Issue found by Sonar

strehle committed:
validate user input

strehle committed:
refactor

strehle committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into fix/regexInCount

coolgang123 commented on server/src/main/java/org/cloudfoundry/identity/uaa/resources/jdbc/JdbcPagingList.java:
```diff
@@ -92,20 +94,23 @@ public E get(int index) {
     private String getCountSql(String sql) {
-        String result = sql.replaceAll("(?i)select (.*?) from (.*)", "select count(*) from $2");
-        int orderByPos = result.toLowerCase().lastIndexOf("order by");
+        String result = UaaStringUtils.getValidatedString(sql).toLowerCase();
```
Converting it to lower case will not work as intended. Consider the sql query with a where clause with upper case criteria:

SELECT Foo.id, FOO.NAME from foo where name = 'Bar'

strehle replied:
Thanks for the hint

coolgang123 commented on server/src/test/java/org/cloudfoundry/identity/uaa/resources/jdbc/JdbcPagingListTests.java:
```diff
@@ -105,6 +106,23 @@ void jumpOverPages() {
     }
 
+    @Test
+    void selectColumnsFull() {
```
Could you add a test case with where and orderBy please

strehle committed:
Merge remote-tracking branch 'origin/develop' into fix/regexInCount

strehle committed:
Review:
Move toLower only compare methods
Add a more complex test with mix cases in SQL using where and order by

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/resources/jdbc/JdbcPagingList.java:
```diff
@@ -73,7 +74,7 @@
         this.args = args;
         this.mapper = mapper;
-        this.size = parameterJdbcTemplate.queryForObject(getCountSql(sql), args, Integer.class);
+        this.size = Optional.ofNullable(parameterJdbcTemplate.queryForObject(getCountSql(sql), args, Integer.class)).orElse(0);
```
SQLs are not from users but internally

strehle committed:
test fix
